File: datasets/build_unified_team_match_panel.py
"""
build_unified_team_match_panel.py
-----------------------------------
Pools the WC2026 team-match panel (ESPN box scores, our own data) with the
WC2018 + WC2022 team-match panel (StatsBomb event-derived counts) into a
single flat CSV, one row per team per match, tagged by season and source.

READ-ONLY against:
  data/processed/espn_match_panel.csv
  data/processed/statsbomb_team_match_panel.csv
Writes only:
  data/processed/unified_team_match_panel.csv

IMPORTANT CAVEAT — measurement heterogeneity, not just schema alignment:
ESPN's per-match stats are a vendor box-score aggregation; StatsBomb's are
derived directly from tagged, timestamped events. These are NOT guaranteed to
count the same thing the same way (this project already hit exactly this
problem once with FBref internationals — see TRAINING_DATASET_STRATEGY doc
Section 4 — "not available"/inconsistent stats were the reason FBref was
rejected in favor of StatsBomb for the historical side). The `data_source`
column exists specifically so nothing downstream silently pools across
sources without knowing it crossed that boundary. A same-team cross-source
rate comparison is printed at the end of this script as a lightweight
plausibility check (not a proof of equivalence — no WC2026 match exists in
both sources to directly cross-validate against).

Column alignment:
  Present in both, aligned name: goals, opponent_goals, result, total_shots,
    shots_on_target, blocked_shots, fouls_committed, yellow_cards, red_cards,
    corners, offsides, passes, pass_pct, clearances, interceptions
  StatsBomb-only (null for ESPN/2026 rows): xg_total, fouls_won,
    passes_completed, referee_name, competition_stage
  ESPN-only (null for StatsBomb/2018+2022 rows): possession_pct, saves,
    effective_tackles, venue_city

Usage: python3 datasets/build_unified_team_match_panel.py
"""

import csv
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT / "datasets"))
from build_master_dataset import CODE_TO_NAME  # reuse existing canonical map, don't duplicate it

ESPN_PANEL = ROOT / "data" / "processed" / "espn_match_panel.csv"
SB_PANEL = ROOT / "data" / "processed" / "statsbomb_team_match_panel.csv"
OUT = ROOT / "data" / "processed" / "unified_team_match_panel.csv"

NAME_TO_CODE = {v: k for k, v in CODE_TO_NAME.items()}

COLUMNS = [
    "season_year", "data_source", "match_id", "match_date", "competition_stage",
    "team_code", "team_name", "opponent_code", "opponent_name", "is_home",
    "goals", "opponent_goals", "result",
    "total_shots", "shots_on_target", "blocked_shots", "xg_total",
    "fouls_committed", "fouls_won", "yellow_cards", "red_cards",
    "corners", "offsides", "passes", "passes_completed", "pass_pct",
    "clearances", "interceptions", "possession_pct", "saves", "effective_tackles",
    "venue", "venue_city", "referee_name",
]


def none_if_blank(v):
    return v if v not in (None, "") else None


def load_espn_rows():
    rows = []
    with open(ESPN_PANEL) as f:
        for r in csv.DictReader(f):
            slug_codes = r["match_slug"].split("-")
            opp_code = slug_codes[1] if slug_codes[0] == r["team_code"] else slug_codes[0]
            goals, opp_goals = int(r["goals"]), int(r["opponent_goals"])
            result = "W" if goals > opp_goals else ("D" if goals == opp_goals else "L")
            pass_pct = float(r["pass_pct"]) * 100 if r["pass_pct"] else None  # ESPN stores as 0-1 fraction
            rows.append({
                "season_year": 2026, "data_source": "espn_boxscore",
                "match_id": r["event_id"], "match_date": r["date"], "competition_stage": None,
                "team_code": r["team_code"], "team_name": CODE_TO_NAME.get(r["team_code"], r["team_code"]),
                "opponent_code": opp_code, "opponent_name": CODE_TO_NAME.get(opp_code, opp_code),
                "is_home": r["home_away"] == "home",
                "goals": goals, "opponent_goals": opp_goals, "result": result,
                "total_shots": r["total_shots"], "shots_on_target": r["shots_on_target"],
                "blocked_shots": r["blocked_shots"], "xg_total": None,
                "fouls_committed": r["fouls"], "fouls_won": None,
                "yellow_cards": r["yellow_cards"], "red_cards": r["red_cards"],
                "corners": r["corners"], "offsides": r["offsides"],
                "passes": r["passes"], "passes_completed": None, "pass_pct": pass_pct,
                "clearances": r["clearances"], "interceptions": r["interceptions"],
                "possession_pct": r["possession_pct"], "saves": r["saves"],
                "effective_tackles": r["effective_tackles"],
                "venue": r["venue"], "venue_city": r["venue_city"], "referee_name": None,
            })
    return rows


def load_statsbomb_rows():
    rows = []
    with open(SB_PANEL) as f:
        for r in csv.DictReader(f):
            year = int(r["season_name"])
            rows.append({
                "season_year": year, "data_source": "statsbomb_event_data",
                "match_id": r["match_id"], "match_date": r["match_date"],
                "competition_stage": r["competition_stage"],
                "team_code": NAME_TO_CODE.get(r["team_name"]), "team_name": r["team_name"],
                "opponent_code": NAME_TO_CODE.get(r["opponent_name"]), "opponent_name": r["opponent_name"],
                "is_home": r["is_home"],
                "goals": r["goals"], "opponent_goals": r["opponent_goals"], "result": r["result"],
                "total_shots": r["total_shots"], "shots_on_target": r["shots_on_target"],
                "blocked_shots": r["blocked_shots"], "xg_total": r["xg_total"],
                "fouls_committed": r["fouls_committed"], "fouls_won": r["fouls_won"],
                "yellow_cards": r["yellow_cards"], "red_cards": r["red_cards"],
                "corners": r["corners"], "offsides": r["offsides"],
                "passes": r["passes"], "passes_completed": r["passes_completed"], "pass_pct": r["pass_pct"],
                "clearances": r["clearances"], "interceptions": r["interceptions"],
                "possession_pct": None, "saves": None, "effective_tackles": None,
                "venue": r["stadium_name"], "venue_city": None, "referee_name": r["referee_name"],
            })
    return rows


def cross_source_plausibility_check(rows):
    """For teams appearing under BOTH data sources (different tournaments,
    since no WC2026 match exists in the StatsBomb set), compare per-match
    rates as a rough plausibility check -- NOT proof the sources measure
    identically, just a flag if something is wildly off scale."""
    by_team_source = {}
    for r in rows:
        key = (r["team_name"], r["data_source"])
        by_team_source.setdefault(key, []).append(r)

    teams_in_both = sorted({t for (t, s) in by_team_source if s == "espn_boxscore"}
                            & {t for (t, s) in by_team_source if s == "statsbomb_event_data"})

    print(f"\n{len(teams_in_both)} teams have rows under both sources (cross-tournament, not cross-validating the same match):")
    print(f"{'Team':<15} {'src':<10} {'n':>3} {'fouls':>7} {'SOT':>6} {'cards':>6} {'corners':>8}")
    for team in teams_in_both:
        for source in ("espn_boxscore", "statsbomb_event_data"):
            trows = by_team_source[(team, source)]
            n = len(trows)
            fouls = sum(float(r["fouls_committed"]) for r in trows) / n
            sot = sum(float(r["shots_on_target"]) for r in trows) / n
            cards = sum(float(r["yellow_cards"]) + float(r["red_cards"]) for r in trows) / n
            corners = sum(float(r["corners"]) for r in trows) / n
            tag = "ESPN/26" if source == "espn_boxscore" else "SB/18+22"
            print(f"{team:<15} {tag:<10} {n:>3} {fouls:>7.2f} {sot:>6.2f} {cards:>6.2f} {corners:>8.2f}")


def build():
    rows = load_espn_rows() + load_statsbomb_rows()
    OUT.parent.mkdir(parents=True, exist_ok=True)
    with open(OUT, "w", newline="") as f:
        w = csv.DictWriter(f, fieldnames=COLUMNS)
        w.writeheader()
        for r in rows:
            w.writerow({c: none_if_blank(r.get(c)) for c in COLUMNS})

    by_year = {}
    for r in rows:
        by_year.setdefault(r["season_year"], 0)
        by_year[r["season_year"]] += 1
    print(f"Unified panel written: {OUT}")
    print(f"Total rows: {len(rows)}")
    for y in sorted(by_year):
        print(f"  {y}: {by_year[y]} team-match rows")

    cross_source_plausibility_check(rows)


if __name__ == "__main__":
    build()

File: datasets/build_unified_team_match_panel.R
#!/usr/bin/env Rscript
# build_unified_team_match_panel.R
# -----------------------------------
# R port of build_unified_team_match_panel.py. Pools the WC2026 ESPN
# team-match panel with the WC2018+2022 StatsBomb team-match panel into one
# tagged table, and reports the same cross-source measurement-heterogeneity
# check (see the .py docstring / STATSBOMB_INTEGRATION_AND_STATS_TESTS doc
# for the caveat this check exists to surface).
#
# Reads (read-only): data/processed/espn_match_panel.csv,
#                     data/processed/statsbomb_team_match_panel.csv
# Writes: data/processed/unified_team_match_panel_r.csv
#
# Usage: Rscript datasets/build_unified_team_match_panel.R

suppressPackageStartupMessages(library(dplyr))

ROOT <- getwd()
ESPN_PANEL <- file.path(ROOT, "data", "processed", "espn_match_panel.csv")
SB_PANEL <- file.path(ROOT, "data", "processed", "statsbomb_team_match_panel.csv")
OUT <- file.path(ROOT, "data", "processed", "unified_team_match_panel_r.csv")

# Reuses the same canonical map as datasets/build_master_dataset.py (kept as
# a literal copy here since sourcing the Python file isn't possible from R;
# keep in sync if that file's CODE_TO_NAME ever changes).
CODE_TO_NAME <- c(
  ALG = "Algeria", ARG = "Argentina", AUS = "Australia", AUT = "Austria",
  BEL = "Belgium", BIH = "Bosnia and Herzegovina", BRA = "Brazil",
  CAN = "Canada", CDR = "DR Congo", CIV = "Ivory Coast", COL = "Colombia",
  CPV = "Cape Verde", CRO = "Croatia", CUR = "Curaçao", CZE = "Czech Republic",
  ECU = "Ecuador", EGY = "Egypt", ENG = "England", ESP = "Spain",
  FRA = "France", GER = "Germany", GHA = "Ghana", HAI = "Haiti",
  IRN = "Iran", IRQ = "Iraq", JOR = "Jordan", JPN = "Japan",
  KOR = "South Korea", KSA = "Saudi Arabia", MAR = "Morocco", MEX = "Mexico",
  NED = "Netherlands", NOR = "Norway", NZL = "New Zealand", PAN = "Panama",
  PAR = "Paraguay", POR = "Portugal", QAT = "Qatar", RSA = "South Africa",
  SAU = "Saudi Arabia", SCO = "Scotland", SEN = "Senegal", SUI = "Switzerland",
  SWE = "Sweden", TUN = "Tunisia", TUR = "Turkey", URU = "Uruguay",
  USA = "United States", UZB = "Uzbekistan"
)
NAME_TO_CODE <- setNames(names(CODE_TO_NAME), CODE_TO_NAME)

load_espn_rows <- function() {
  r <- read.csv(ESPN_PANEL, stringsAsFactors = FALSE)
  slug_a <- sapply(strsplit(r$match_slug, "-"), `[`, 1)
  slug_b <- sapply(strsplit(r$match_slug, "-"), `[`, 2)
  opp_code <- ifelse(slug_a == r$team_code, slug_b, slug_a)
  result <- ifelse(r$goals > r$opponent_goals, "W", ifelse(r$goals == r$opponent_goals, "D", "L"))

  data.frame(
    season_year = 2026, data_source = "espn_boxscore",
    match_id = r$event_id, match_date = r$date, competition_stage = NA,
    team_code = r$team_code, team_name = unname(CODE_TO_NAME[r$team_code]),
    opponent_code = opp_code, opponent_name = unname(CODE_TO_NAME[opp_code]),
    is_home = r$home_away == "home",
    goals = r$goals, opponent_goals = r$opponent_goals, result = result,
    total_shots = r$total_shots, shots_on_target = r$shots_on_target,
    blocked_shots = r$blocked_shots, xg_total = NA,
    fouls_committed = r$fouls, fouls_won = NA,
    yellow_cards = r$yellow_cards, red_cards = r$red_cards,
    corners = r$corners, offsides = r$offsides,
    passes = r$passes, passes_completed = NA, pass_pct = r$pass_pct * 100,
    clearances = r$clearances, interceptions = r$interceptions,
    possession_pct = r$possession_pct, saves = r$saves,
    effective_tackles = r$effective_tackles,
    venue = r$venue, venue_city = r$venue_city, referee_name = NA,
    stringsAsFactors = FALSE
  )
}

load_statsbomb_rows <- function() {
  r <- read.csv(SB_PANEL, stringsAsFactors = FALSE)
  data.frame(
    season_year = as.integer(r$season_name), data_source = "statsbomb_event_data",
    match_id = r$match_id, match_date = r$match_date, competition_stage = r$competition_stage,
    team_code = unname(NAME_TO_CODE[r$team_name]), team_name = r$team_name,
    opponent_code = unname(NAME_TO_CODE[r$opponent_name]), opponent_name = r$opponent_name,
    is_home = as.logical(r$is_home),
    goals = r$goals, opponent_goals = r$opponent_goals, result = r$result,
    total_shots = r$total_shots, shots_on_target = r$shots_on_target,
    blocked_shots = r$blocked_shots, xg_total = r$xg_total,
    fouls_committed = r$fouls_committed, fouls_won = r$fouls_won,
    yellow_cards = r$yellow_cards, red_cards = r$red_cards,
    corners = r$corners, offsides = r$offsides,
    passes = r$passes, passes_completed = r$passes_completed, pass_pct = r$pass_pct,
    clearances = r$clearances, interceptions = r$interceptions,
    possession_pct = NA, saves = NA, effective_tackles = NA,
    venue = r$stadium_name, venue_city = NA, referee_name = r$referee_name,
    stringsAsFactors = FALSE
  )
}

cross_source_plausibility_check <- function(df) {
  teams_espn <- unique(df$team_name[df$data_source == "espn_boxscore"])
  teams_sb <- unique(df$team_name[df$data_source == "statsbomb_event_data"])
  teams_both <- sort(intersect(teams_espn, teams_sb))

  cat("\n", length(teams_both), "teams appear under both sources (cross-tournament):\n")
  cat(sprintf("%-15s %-10s %3s %7s %6s %6s %8s\n", "Team", "src", "n", "fouls", "SOT", "cards", "corners"))
  ratios <- list(fouls = c(), sot = c(), cards = c(), corners = c())
  for (team in teams_both) {
    for (src in c("espn_boxscore", "statsbomb_event_data")) {
      tr <- df[df$team_name == team & df$data_source == src, ]
      n <- nrow(tr)
      fouls <- mean(tr$fouls_committed); sot <- mean(tr$shots_on_target)
      cards <- mean(tr$yellow_cards + tr$red_cards); corners <- mean(tr$corners)
      tag <- if (src == "espn_boxscore") "ESPN/26" else "SB/18+22"
      cat(sprintf("%-15s %-10s %3d %7.2f %6.2f %6.2f %8.2f\n", team, tag, n, fouls, sot, cards, corners))
    }
    e <- df[df$team_name == team & df$data_source == "espn_boxscore", ]
    s <- df[df$team_name == team & df$data_source == "statsbomb_event_data", ]
    ratios$fouls <- c(ratios$fouls, mean(e$fouls_committed) / mean(s$fouls_committed))
    ratios$sot <- c(ratios$sot, mean(e$shots_on_target) / mean(s$shots_on_target))
    ratios$cards <- c(ratios$cards, mean(e$yellow_cards + e$red_cards) / mean(s$yellow_cards + s$red_cards))
    ratios$corners <- c(ratios$corners, mean(e$corners) / mean(s$corners))
  }
  cat("\nMean ESPN/StatsBomb ratio across", length(teams_both), "teams:\n")
  for (stat in names(ratios)) cat(sprintf("  %-10s %.3f\n", stat, mean(ratios[[stat]])))
}

build <- function() {
  df <- bind_rows(load_espn_rows(), load_statsbomb_rows())
  dir.create(dirname(OUT), recursive = TRUE, showWarnings = FALSE)
  write.csv(df, OUT, row.names = FALSE, na = "")

  cat("Unified panel written:", OUT, "\n")
  cat("Total rows:", nrow(df), "\n")
  for (y in sort(unique(df$season_year))) {
    cat(sprintf("  %d: %d team-match rows\n", y, sum(df$season_year == y)))
  }
  cross_source_plausibility_check(df)
}

build()
